compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/Preprocessing/LimitScatterBounds.cpp

```diff
@@ -0,0 +1,280 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// Implements logic for lowering StableHLO einsum op to dot_general ops.
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "stablehlo-iree/Conversion/Preprocessing/Passes.h"
+#include "stablehlo-iree/Conversion/Preprocessing/Rewriters.h"
+#include "stablehlo/dialect/StablehloOps.h"
+
+namespace mlir::iree_compiler::stablehlo {
+
+#define GEN_PASS_DEF_LIMITSCATTERBOUNDS
+#include "stablehlo-iree/Conversion/Preprocessing/Passes.h.inc"
+
+namespace {
+
+LogicalResult replaceScatter(mlir::stablehlo::ScatterOp op) {
+  IRRewriter builder(op.getContext());
+  ImplicitLocOpBuilder b(op.getLoc(), builder);
+  b.setInsertionPoint(op);
+
+  auto dest = op.getInputs().front();
+  auto indices = op.getScatterIndices();
+  auto update0 = op.getUpdates().front();
+
+  auto destTy = cast<ShapedType>(dest.getType());
+  auto indicesTy = cast<ShapedType>(indices.getType());
+  auto update0Ty = cast<ShapedType>(update0.getType());
+
+  auto indicesETy = indicesTy.getElementType();
+  auto indicesShape = indicesTy.getShape();
+
+  auto dimnums = op.getScatterDimensionNumbers();
+  const int64_t rank = destTy.getRank();
+
+  int64_t indexVectorDim = dimnums.getIndexVectorDim();
+
+  // Determine the bound limits for each index:
+  auto updateWindowDims = dimnums.getUpdateWindowDims();
+  auto toOperandDims = dimnums.getScatterDimsToOperandDims();
+  auto insertedWindowDims = dimnums.getInsertedWindowDims();
+
+  llvm::SmallVector<bool> isUpdatedWindowDim(rank, true);
+  for (auto dim : insertedWindowDims) {
+    isUpdatedWindowDim[dim] = false;
+  }
+
+  // This is grabbing all of the update shapes with indices removed:
+  llvm::SmallVector<Value> updateWindowShape;
+  for (auto dim : updateWindowDims) {
+    updateWindowShape.push_back(b.create<tensor::DimOp>(update0, dim));
+  }
+
+  Value oneIdx = b.create<arith::ConstantIndexOp>(1);
+
+  // We subtract the width of the update for each index. For inserted dimensions
+  // this is 1.
+  llvm::SmallVector<Value> windowBounds;
+  {
+    int j = 0;
+    for (auto it : llvm::enumerate(isUpdatedWindowDim)) {
```

Comment:
Drive-by nit
```suggestion
    for (auto [idx, val] : llvm::enumerate(isUpdatedWindowDim)) {
```

also in other loops below

Reply:
Integrated.